### PR TITLE
DSPDC-498 Add method to list paths in GCS

### PR DESCRIPTION
**Pull request by danxmoran**
I split out a `common` subproject to hold our enum for file types

### CODE REVIEW

File: gcs/src/main/scala/org/broadinstitute/monster/storage/gcs/GcsApi.scala [hunk 1]
```diff
@@ -451,6 +449,89 @@ class GcsApi private[gcs] (runHttp: Request[IO] => Resource[IO, Response[IO]]) {
...
+    * @param bucket name of the bucket to list contents within
+    * @param path prefix to match listed contents against
+    * @param pageSize max number of results to pull from GCS per HTTP request
+    */
+  def listContents(
+    bucket: String,
+    path: String,
+    pageSize: Int
+  ): Stream[IO, (String, FileType)] =
+    // State-tracking is non-obvious here.
```
Comment (danxmoran) [inline gcs/src/main/scala/org/broadinstitute/monster/storage/gcs/GcsApi.scala]: Note for anyone looking at this now: I just pushed a change which doesn't abuse `Option` for state-tracking. It should be simpler to read and review.

File: gcs/src/it/scala/org/broadinstitute/monster/storage/gcs/GcsApiIntegrationSpec.scala [hunk 2]
```diff
@@ -313,7 +345,8 @@ class GcsApiIntegrationSpec
         .chunk(bodyChunk)
         .covary[IO]
         .through(fs2.compress.gzip(1024 * 1024))
-        .drop(10).take(40)
+        .drop(10)
```
Comment (benjamincarlin) [inline gcs/src/it/scala/org/broadinstitute/monster/storage/gcs/GcsApiIntegrationSpec.scala:348]: is there a slice or range?

File: common/src/main/scala/org/broadinstitute/monster/storage/common/OperationStatus.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.broadinstitute.monster.storage.common
+
+import enumeratum.{Enum, EnumEntry}
+
+import scala.collection.immutable.IndexedSeq
+
+private[storage] sealed trait OperationStatus extends EnumEntry
```
Comment (danxmoran) [inline common/src/main/scala/org/broadinstitute/monster/storage/common/OperationStatus.scala:13]: Note to self: needs scaladoc

File: gcs/src/test/scala/org/broadinstitute/monster/storage/gcs/GcsApiSpec.scala [hunk 4]
```diff
@@ -590,4 +597,139 @@ class GcsApiSpec extends FlatSpec with Matchers with OptionValues with EitherVal
...
+      Resource.pure(Response[IO](status = Status.Ok, body = Stream.emits(body.getBytes)))
+    })
+
+    val resultsOrErr =
+      api.listContents(bucket, path, smallPageSize).compile.drain.attempt.unsafeRunSync()
+    resultsOrErr.left.value.getMessage should include(path)
+  }
+
+  it should "raise a helpful error if GCS returns an error code to a list request" in {
+    val body = "OH NO"
```
Comment (benjamincarlin) [inline gcs/src/test/scala/org/broadinstitute/monster/storage/gcs/GcsApiSpec.scala:717]: classic!

**Review by benjamincarlin — APPROVED** [review]

**Review by raaidbroad — APPROVED** [review]

File: gcs/src/it/scala/org/broadinstitute/monster/storage/gcs/GcsApiIntegrationSpec.scala (hunk 2 above)
Comment (danxmoran) [reply gcs/src/it/scala/org/broadinstitute/monster/storage/gcs/GcsApiIntegrationSpec.scala:348]: Not on fs2's Stream

**Pull request merged**